Add db replication blog post (#3753)
pulumi/pulumi-hugo: themes/default/content/blog/replicating-data-across-multi-region-apps/index.md

---

> title: "Replicating Data Across Multi Region Apps"

Reviewer: ```suggestion
title: "Replicating Data to Support Multi-Region Applications"
```

date: 2023-12-15T11:03:10Z
draft: false

> meta_desc: Learn about the tradeoffs in distributed databases when building for multiple regions.

Reviewer: ```suggestion
meta_desc: Learn about the tradeoffs in distributed databases and data replication when using them for multi-region applications.
```

> meta_image: meta.png

Reviewer: Adding a reminder here that we can't publish this until you have a meta image for the post.

authors:
    - adora-nwodo
tags:
    - multi-region-databases
    - data-replication
    - cloud-deployment
    - distributed-systems
    - geographic-distribution
    - high-availability

---

> In [the previous article](https://www.pulumi.com/blog/scaling-apps-across-multiple-regions/), we covered multi-region scaling, its importance, and how you can use Pulumi stacks to represent multiple regions and environments. The big takeaway from that article is that scaling your application across multiple regions is an important architectural decision to enable scalability and availability, but it comes with its own set of considerations. One of these considerations is around the data that your application needs or uses. Data replication is typically necessary in multi-region architectures because if you have multiple application instances worldwide making calls to just one database instance, the latency of your application will be high. After all, it will take a long time for that database instance to perform data operations and send back a result to the user through the application. Some requests might have repeated timeouts. In this article, we will cover data replication in multi-region applications and how it plays a pivotal role in distributed systems.

Reviewer: For all links to pages on the Pulumi site, please use a relative link (`/blog/scaling-apps-across-multiple-regions` instead of `https://www.pulumi.com/blog/...`). I'll comment on this only once here, but please ensure the change is made throughout the entire PR.

As mentioned earlier, though, more than just network partitioning comes into play here. Network bandwidth affects the balance between preserving consistency or optimizing for availability in the context of multi-region applications. Network configurations also come into play; what impact will a larger MTU (Maximum Transmission Unit) have on data replication? Decisions here are full of trade-offs. For example, prioritizing replication speed to ensure data consistency across regions may lead to increased latency, affecting the system's ability to meet stringent availability requirements. An extension to the CAP theorem, known as [PACELC](https://en.wikipedia.org/wiki/PACELC_theorem), captures the trade-off between latency and consistency even when network partitions aren’t present.

Therefore, the challenge lies in finding the right balance. Striking a compromise between data consistency and system performance (in the form of availability) is a delicate task requiring a careful analysis of the principles of the CAP theorem, guiding system architects in making informed choices that align with the specific needs and priorities of their distributed applications.

> ## Using Pulumi for data replication in your cloud service

Reviewer: ```suggestion
## Using Pulumi to configure data replication
```

Here’s how to configure CosmosDB to prioritize availability:

```typescript
const cosmosDbAccount = new azure.documentdb.DatabaseAccount("myCosmosDbAccount", {
    resourceGroupName: resourceGroup.name,
    location: resourceGroup.location,
    locations: [
        { locationName: "East US" },
        { locationName: "West US" },
        // Add additional Azure regions for replication
    ],
    consistencyPolicy: {
        defaultConsistencyLevel: "Eventual",
    },
    enableAutomaticFailover: true,
});
```

When prioritizing availability, you might choose to use a default consistency level of “Eventual,” as shown in the example above. Eventual consistency prioritizes high availability by enabling updates to propagate gradually across replicas. While it has decreased read latency, it accepts temporary inconsistencies that resolve over time. As you might expect, your application will need to be prepared to deal with these temporary inconsistencies.

> A more balanced option that still prioritizes availability is called “BoundedStaleness.” This choice introduces staleness in read operations, allowing for better availability during network partitions. "BoundedStaleness" provides a controlled staleness window (maxStalenessPrefix), enabling developers to fine-tune the balance between consistency and availability. Bounded Staleness offers a sort of middle ground, allowing users to define a controlled maximum lag for read operations and therefore ensuring a specified level of consistency while maintaining availability during network partitions. This model suits scenarios where a compromise between consistency and availability is necessary, and the application can tolerate a defined level of data staleness.

Reviewer: ```suggestion
A more balanced option is called “BoundedStaleness.” This choice introduces staleness in read operations, allowing for better availability during network partitions. "BoundedStaleness" provides a controlled staleness window (maxStalenessPrefix), enabling developers to fine-tune the balance between consistency and availability. Bounded Staleness offers a sort of middle ground, allowing users to define a controlled maximum lag for read operations and therefore ensuring a specified level of consistency while maintaining availability during network partitions. This model suits scenarios where a compromise between consistency and availability is necessary, and the application can tolerate a defined level of data staleness.
```
